### ISSUE TITLE
Returning / accepting tuples

### ISSUE THREAD

**Issue by shepmaster** [shepmaster/rust-ffi-omnibus#9]
```rust
fn flip_things_around(tup: (u32, u32)) -> (u32, u32) {
    let (a, b) = tup;
    (a-1, b+1)
}
```

**Comment by flo-l** [comment]
I think in ruby the ffi gem can handle tuples via structs, see [my comment](https://github.com/shepmaster/rust-ffi-omnibus/issues/2#issuecomment-116225015) on #2 for code!

You can choose how the struct should be passed/returned by using ```FFI::Struct#by_value``` or ```FFI::Struct#by_ref```

**Comment by shepmaster** [comment]
@flo-l absolutely! Tuples [are just structs](http://stackoverflow.com/a/30992210/155423), but the compiler is quite helpful in doing a lot of the implementation for us!

**Comment by shepmaster** [comment]
@aisamanra Based on [your example repo](https://github.com/aisamanra/rust-haskell-ffi/blob/e7114da60484fad7fabcd72885cf9de9bd7c1f7e/point.rs), and [this Stack Overflow question](http://stackoverflow.com/q/10903940/155423), am I correct in stating that Haskell cannot directly take or return a structure, and it must always be a pointer to the struct?

**Comment by shepmaster** [comment]
Oh, and I guess [this proposal to support structs](https://ghc.haskell.org/trac/ghc/wiki/CStructures) in Haskell FFI too. :-(

**Comment by aisamanra** [comment]
Sadly, yes. There is a weird, obscure way of working with structs that aren't behind a pointer, but it only works with foreign functions that use a peculiar GHC-specific calling convention, and making that happen is [effectively sorcery](http://breaks.for.alienz.org/blog/2012/02/09/parsing-market-data-feeds-with-ragel/). (I've thought about doing this for Rust code interop, just as a proof-of-concept, and then documenting it, but I haven't actually tried yet. I really ought to get around to it.)

I think the solution might be to mention, "For all intents and purposes, in Haskell you can only do this when there's a layer of indirection," and then include the equivalent code for `&(u32, u32)`.

**Comment by shepmaster** [comment]
Closed via 22b65627918ac3164a492e7eb45cc809b74941eb
